todoapp/app/src/test/java/com/example/android/architecture/blueprints/todoapp/taskdetail/domain/LogicTest.kt

```diff
@@ -0,0 +1,91 @@
+        spec.given(activeTask)
+                .whenEvent(TaskDetailEvent.CompleteTaskRequested)
+                .then(
+                        assertThatNext(
+                                hasModel(activeTask.complete()),
+                                hasEffects(TaskDetailEffect.SaveTask(activeTask.complete()) as TaskDetailEffect))
+                )
+    }
+
+    @Test
+    fun activatingATaskUncompletesAndSavesIt() {
+        val details = TaskDetails("T1", "description", true)
+        val completedTask = Task("123", details)
+        spec.given(completedTask)
+                .whenEvent(TaskDetailEvent.ActivateTaskRequested)
+                .then(
+                        assertThatNext(
+                                hasModel(completedTask.activate()),
+                                hasEffects(TaskDetailEffect.SaveTask(completedTask.activate()) as TaskDetailEffect)))
+    }
+
+    @Test
+    fun completingACompletedTaskDoesNothing() {
+        val details = TaskDetails("T1", "description", true)
+        val completedTask = Task("123", details)
+        spec.given(completedTask).whenEvent(TaskDetailEvent.CompleteTaskRequested).then(assertThatNext(hasNothing()))
+    }
+
+    @Test
+    fun activatingAnActiveTaskDoesNothing() {
+        val activeTask = createTask()
+        spec.given(activeTask).`when`(TaskDetailEvent.ActivateTaskRequested).then(assertThatNext(hasNothing()))
+    }
+
+    @Test
+    fun editingATaskWillOpenTaskEditor() {
+        val someTask = createTask()
+        spec.given(someTask)
+                .whenEvent(TaskDetailEvent.EditTaskRequested)
+                .then(assertThatNext(hasEffects(TaskDetailEffect.OpenTaskEditor(someTask) as TaskDetailEffect)))
+    }
+
+    @Test
+    fun requestingTaskDeletionDeletesTheTask() {
+        val someTask = createTask()
+        spec.given(someTask)
+                .whenEvent(TaskDetailEvent.DeleteTaskRequested)
+                .then(assertThatNext(hasEffects(TaskDetailEffect.DeleteTask(someTask) as TaskDetailEffect)))
+    }
+
+    @Test
+    fun taskDeletionExitsDetailsFeature() {
+        val someTask = createTask()
+        spec.given(someTask).whenEvent(TaskDetailEvent.TaskDeleted).then(assertThatNext(hasEffects(TaskDetailEffect.Exit as TaskDetailEffect)))
+    }
+
+    @Test
+    fun taskCompletionShowsFeedback() {
+        val someTask = createTask()
+        spec.given(someTask)
+                .whenEvent(TaskDetailEvent.TaskMarkedComplete)
+                .then(assertThatNext(hasEffects(TaskDetailEffect.NotifyTaskMarkedComplete as TaskDetailEffect)))
+    }
+
+    @Test
+    fun taskUpdateShowsFeedback() {
+        val someTask = createTask()
+        spec.given(someTask)
+                .whenEvent(TaskDetailEvent.TaskMarkedActive)
+                .then(assertThatNext(hasEffects(TaskDetailEffect.NotifyTaskMarkedActive as TaskDetailEffect)))
+    }
+
+
+    private fun createTask() = Task("123", TaskDetails("T1", "description"))
```

Comment:
This function returns an immutable value. Just declare it as a value instead and you won't need to re-instantiate it everywhere.

Reply:
I created a val named `someTask` and used that instead.